[Firrtl] Store port names more efficiently as module attribute instead of argument attribute

prithayan opened the pull request:
Replace the `firrtl.name` attribute that exists on individual arguments with `portNames` attribute on the module declaration. 
Store the names as a separate attribute with a single array attribute.

Replace the following::
```firrtl.extmodule @ExternalModule(!firrtl.bundle<valid: uint<1>, ready: flip<uint<1>>, data: uint<64>> {firrtl.name = "source"})```
With this ::
```firrtl.extmodule @ExternalModule(!firrtl.bundle<valid: uint<1>, ready: flip<uint<1>>, data: uint<64>> ) attributes {argNames = ["source"]}```

It allows directly indexing port names, make the # entries in the module dictionary constant time. 

This is the FIRRTL hallf of: https://github.com/llvm/circt/issues/872

youngar commented on lib/Dialect/FIRRTL/FIRRTLDialect.cpp:
```diff
@@ -21,19 +21,15 @@ using namespace firrtl;
+StringAttr firrtl::getFIRRTLModuleArgNameAttr(Operation *module, size_t argNo) {
+  if (!hasFIRRTLModuleArgNameAttr(module))
+    return StringAttr();
+  return module->getAttrOfType<ArrayAttr>("argNames")[argNo].cast<StringAttr>();
```
I would optimize this to avoid looking up the argument array each time

prithayan replied:
Thanks for the suggestion, I updated this to single lookup.

seldridge commented on test/Dialect/FIRRTL/lower-types.mlir:
```diff
@@ -89,7 +89,8 @@ firrtl.circuit "Uniquification" {
   // CHECK-NOT: %[[FLATTENED_ARG]]
-  // CHECK-SAME: %[[RENAMED_ARG:a_b.+]]: [[RENAMED_TYPE:!firrtl.uint<1>]] {firrtl.name = "[[FLATTENED_ARG]]"}
+  // CHECK-SAME: %[[RENAMED_ARG:a_b.+]]: [[RENAMED_TYPE:!firrtl.uint<1>]] 
+  // CHECK-SAME: {argNames = ["a_b", "a_b"]}
```
This behavior seems weird, but it's the old behavior..

seldridge commented on include/circt/Dialect/FIRRTL/FIRRTLDialect.h:
```diff
@@ -22,10 +22,18 @@ namespace firrtl {
+                                               size_t argNo) {
+  return hasFIRRTLModuleArgNameAttr(module)
+             ? getFIRRTLModuleArgNameAttr(module, argNo).getValue()
+             : "";
```
Is it safe to return `""` as a `StringRef`? An alternative formulation of this would be to guarantee that this can succeed (there exists a `StringAttr`) and return the ref to that.

seldridge replied:
Also, the change to module construction to pre-populate everything with `StringAttr("")` for empty names should mean that false path can never happen?

seldridge reviewed (commented):
High level: (1) looking good, (2) output looks fantastic. I hadn't couldn't on the reduction in textual IR size.

I think there are some superfluous checks that the arg name exists and that a port has a name. I think the change to module construction should guarantee that this exists and some of the code will simplify a bit.

youngar commented:
I think that it would make sense to require every module parameter to have a name and check this in the verifier. This means you won't have to check for the arg array before lookup in a bunch of places. (looks like Schuyler beat me to it :))

I am also noticing that when a module has no arguments, we are not attaching an empty argument array, there is no arg attribute at all.  I'm not sure if its better, but we could make the arg array a required attribute, and add it to the list of op arguments in the TD file.  We made a similar change with requiring names, and no name is now "".

prithayan commented:
Thanks @youngar , @seldridge . I think adding the default argNames attribute makes sense. I will work on the suggested changes.

prithayan commented:
Updated to address the review comments, 
1. Add a default ``"argNames"`` attribute, with empty strings in the ``builder``
2. Remove all redundant checks 
3. Updated the API, to get the named attribute once and then index into it.

youngar commented on lib/Dialect/FIRRTL/FIRRTLDialect.cpp:
```diff
@@ -113,9 +103,12 @@ struct FIRRTLOpAsmDialectInterface : public OpAsmDialectInterface {
+      StringAttr str =
+          argAttr[i]
+              .cast<StringAttr>();
+      if (!str.getValue().empty())
```
You can save calling `getValue()` twice:
```c++
      StringRef argName = argAttr[i].cast<StringAttr>().getValue();
```

youngar commented on lib/Dialect/FIRRTL/FIRRTLOps.cpp:
```diff
@@ -253,16 +253,16 @@ static void buildModule(OpBuilder &builder, OperationState &result,
-        NamedAttribute(builder.getIdentifier("firrtl.name"), ports[i].name);
+      argNames.push_back(builder.getStringAttr(""));
+    else
+      argNames.push_back(ports[i].name);
```
I don't think you need to check for an empty name anymore.. 
```
  for (size_t i = 0, e = ports.size(); i != e; ++i)
    argNames.push_back(ports[i].name);
```

youngar replied:
you can resize the argNames array to be the same size as the ports array

prithayan replied:
The port.name might not exist, so I think we would need the empty check, 
```
 struct ModulePortInfo {
... 
   StringRef getName() const { return name ? name.getValue() : ""; } 

```

lattner replied:
The `getName` method should be changed.  We should enforce the invariant that the portNames list is the same size as the # ports in a firrtl module, and always has (possibly empty?) StringAttr members in its array.  This will simplify a bunch of this code and make it less flakey.

youngar commented on lib/Dialect/FIRRTL/FIRRTLOps.cpp:
```diff
@@ -283,16 +283,22 @@ void FModuleOp::build(OpBuilder &builder, OperationState &result,
+  SmallVector<Attribute> argNames;
+  for (size_t i = 0, e = ports.size(); i != e; i++)
+    argNames.push_back(builder.getStringAttr(""));
+  result.addAttribute("argNames", builder.getArrayAttr(argNames));
```
I think we're guaranteed to have argNames and this whole block of code can be removed?

darthscsi replied:
buildModule just created argNames.

youngar commented on lib/Dialect/FIRRTL/FIRRTLOps.cpp:
```diff
@@ -314,47 +320,36 @@ void FExtModuleOp::build(OpBuilder &builder, OperationState &result,
 // allow this customization.
 static void printFunctionSignature2(OpAsmPrinter &p, Operation *op,
```
I think at this point it would make sense to give this function a more meaningful name and remove the TODO

youngar commented on lib/Dialect/FIRRTL/FIRRTLOps.cpp:
```diff
@@ -253,16 +253,16 @@ static void buildModule(OpBuilder &builder, OperationState &result,
-                        builder.getDictionaryAttr(argAttr));
-  }
+  if (!argNames.empty())
+    result.addAttribute("argNames", builder.getArrayAttr(argNames));
```
I guess this wasn't an entirely accurate statement, but I think we should unconditionally add the argNames attribute.  Is there a reason it's structured this way? (e.g. to avoid putting an empty attribute on an external modules?)

darthscsi replied:
port names makes sense for external modules too.  Should the verifier check that argNames exists and is long enough?

darthscsi commented on lib/Dialect/FIRRTL/FIRRTLDialect.cpp:
```diff
@@ -113,9 +103,12 @@ struct FIRRTLOpAsmDialectInterface : public OpAsmDialectInterface {
-      if (auto str = getFIRRTLNameAttr(mlir::impl::getArgAttrs(parentOp, i)))
+      StringAttr str =
+          argAttr[i]
+              .cast<StringAttr>();
```
clang-format?

darthscsi commented on test/Dialect/FIRRTL/annotations.fir:
```diff
@@ -65,7 +65,7 @@ circuit Foo: %[[{"a":"a","target":"~Foo|Bar"}]]
     ; CHECK-LABEL: module {
     ; CHECK: firrtl.extmodule @Bar
-    ; CHECK-SAME: attributes {annotations = [{a = "a"}]}
+    ; CHECK-SAME: attributes {annotations = [{a = "a"}]
```
seems fine to leave the '}'

prithayan commented:
Thanks @youngar , @darthscsi , for the detailed review.
Addressed all the review comments. 
Added the verifier, to check for argNames

seldridge commented on lib/Dialect/FIRRTL/Transforms/LowerTypes.cpp:
```diff
@@ -828,13 +813,16 @@ Value TypeLoweringVisitor::addArg(FModuleOp module, Type type,
+      getFIRRTLModuleArgNameAttr(module)[oldArgNumber].cast<StringAttr>();
+  Attribute newArg;
+  if (!nameAttr.getValue().empty()) {
+    SmallString<16> newName(nameAttr.getValue());
+    newName += nameSuffix;
+    newArg = builder->getStringAttr(newName);
```
You can rely on some `llvm::Twine` magic to avoid the string concatenation until `getStringAttr` actually needs it:
```suggestion
    newArg = builder->getStringAttr(nameAttr.getValue() + nameSuffix);
```

prithayan replied:
Thanks Schuyler, made the changes

seldridge reviewed (approved):
Looks great. I'm loving how this cleans up the MLIR textual representation.

lattner commented:
Hey @prithayan, please let me know when this is settled and I'll do a final review etc.  Thank you for driving this forward!

prithayan commented:
@lattner , please review, Thanks

lattner commented on include/circt/Dialect/FIRRTL/FIRRTLDialect.h:
```diff
@@ -22,9 +22,9 @@ namespace firrtl {
 class FIRRTLType;
 
-/// If the specified attribute list has a firrtl.name attribute, return its
+/// If the specified module has an argnames attribute, return its
```
Nit: modules will always have this, so please make this more direct, e.g. "Return the argNames attribute for the specified module, which contains the name for each port."

I'd also recommend renaming this to `getModulePortNames` since it is in the firrtl namespace.  This also aligns with the `getModulePortName` function naming.

lattner commented on lib/Dialect/FIRRTL/FIRRTLDialect.cpp:
```diff
@@ -21,19 +21,9 @@ using namespace firrtl;
-  return StringAttr();
+// If the specified module contains the argNames attribute, return it.
+ArrayAttr firrtl::getFIRRTLModuleArgNameAttr(Operation *module) {
+  return module->getAttrOfType<ArrayAttr>("argNames");
```
I'd recommend naming the attribute `portNames`.  The RTL dialect splits "arguments" and "results" but in firrtl they are all "ports" which happen to be modeled as arguments.

lattner commented on lib/Dialect/FIRRTL/FIRRTLOps.cpp:
```diff
@@ -526,6 +513,11 @@ static LogicalResult verifyFExtModuleOp(FExtModuleOp op) {
+  auto argAttr = getFIRRTLModuleArgNameAttr(op);
+
+  if (op.getPorts().size() != argAttr.size()) {
+    return failure();
```
Don't these need to emit errors before returning failure?

lattner replied:
This still needs to be something like `return op.emitError(...`, not `return failure()`.

lattner commented on lib/Dialect/FIRRTL/Transforms/LowerTypes.cpp:
```diff
@@ -805,17 +801,6 @@ void TypeLoweringVisitor::visitExpr(InvalidValuePrimOp op) {
 // Helpers to manage state.
 //===----------------------------------------------------------------------===//
 
-static DictionaryAttr getArgAttrs(StringAttr nameAttr, StringRef suffix,
```
I'm so happy to see this go away :-)

lattner commented on lib/Dialect/FIRRTL/Transforms/LowerTypes.cpp:
```diff
@@ -827,13 +812,15 @@ Value TypeLoweringVisitor::addArg(FModuleOp module, Type type,
-    newModuleAttrs.push_back(NamedAttribute(newArgAttrName, newArgAttrs));
-  }
+  StringAttr nameAttr =
+      getFIRRTLModuleArgNameAttr(module)[oldArgNumber].cast<StringAttr>();
```
I think there is a `getModulePortName(module, idx)` method you can use here.

lattner commented on lib/Dialect/FIRRTL/Transforms/LowerTypes.cpp:
```diff
@@ -827,13 +812,15 @@ Value TypeLoweringVisitor::addArg(FModuleOp module, Type type,
+    newArg =
+        builder->getStringAttr(nameAttr.getValue().str() + nameSuffix.str());
+  else
+    newArg = builder->getStringAttr("");
```
Does firrtl dialect actually allow empty port names?  If not, then all this can go away.  If not, we should check that in the verifier.

lattner commented on include/circt/Dialect/FIRRTL/FIRRTLDialect.h:
```diff
@@ -22,9 +22,11 @@ namespace firrtl {
-/// If the specified attribute list has a firrtl.name attribute, return its
-/// value.
-StringAttr getFIRRTLNameAttr(ArrayRef<NamedAttribute> attrs);
+const StringRef portNameAttrStringRef("portNames");
```
We don't typically do this for other attributes.  Please just use the magic constant where necessary.  Thanks.

lattner commented on include/circt/Dialect/FIRRTL/FIRRTLDialect.h:
```diff
@@ -22,9 +22,11 @@ namespace firrtl {
+
+/// Return the portNames attribute for the specified module, which contains the
+/// name for each port.
+ArrayAttr getModulePortNames(Operation *module);
```
Thanks for the rename, plz move this to FIRRTLOps.h to be next to getModulePortName.

lattner commented on lib/Dialect/FIRRTL/FIRRTLOps.cpp:
```diff
@@ -312,47 +311,36 @@ void FExtModuleOp::build(OpBuilder &builder, OperationState &result,
+  SmallString<32> resultNameStr;
 
   p << '(';
+  auto argAttr = getModulePortNames(op);
```
`argAttr` -> `portNamesAttr`?

lattner commented on lib/Dialect/FIRRTL/FIRRTLOps.cpp:
```diff
@@ -443,35 +437,28 @@ static ParseResult parseFModuleOp(OpAsmParser &parser, OperationState &result,
+    // verbosity in dumps of including it explicitly in the attribute
+    // dictionary.
+    for (size_t i = 0, e = entryArgs.size(); i != e; ++i) {
+
```
super nit: no need for the blank line

lattner reviewed (approved):
Awesome, thank you!

prithayan commented:
Thank you @lattner, for the detailed feedback. Made all the requested changes, 

- Rename ``argNames`` attribute to ``portNames``
- Port names can never be empty, remove checks for empty port names
- Move ``getModulePortNames`` from FIRRTLDialect.h to FIRRTLOps.h

prithayan commented:
Thank you all, for the extensive review on this PR. Merged it with main.

lattner commented:
Woohoo great work!
